File: backend/app/schemas/vistas.py
# app/schemas/vistas.py

from pydantic import BaseModel
from typing import Literal, Optional, Union
from datetime import date
from app.models.enums import RolPersonaTipo, GeneroTipo, TipoTarjeta


class PosicionSchema(BaseModel):
    id_torneo: int
    torneo: str
    id_equipo: int
    equipo: str
    partidos_jugados: int
    ganados: int
    empatados: int
    perdidos: int
    goles_a_favor: int
    goles_en_contra: int
    diferencia_gol: int
    puntos: int

    class Config:
        from_attributes = True # Para que funcione bien con ORMs como SQLAlchemy

class PersonaConRol(BaseModel):
    """Schema para la vista vw_persona_roles"""
    id_persona: int
    nombre: str
    apellido: str
    documento: Optional[Union[str, int]] = None  # Acepta string o integer
    genero: GeneroTipo
    rol: RolPersonaTipo
    fecha_desde: Optional[date] = None
    fecha_hasta: Optional[date] = None

    class Config:
        from_attributes = True


# inscripcion_torneo_detalle

class InscripcionTorneoDetalle(BaseModel):
    """ Schema para la vista vw_inscripcion_torneo_detalle """
    id_inscripcion: int
    id_torneo: int
    id_equipo: int

    nombre_equipo: str
    categoria_equipo: str
    genero_equipo: str

    id_club: int
    nombre_club: str

    fecha_inscripcion: date
    fecha_baja: Optional[date] = None

    class Config:
        from_attributes = True

# vista club_personas_roles

class ClubPersonaRolOut(BaseModel):
    id_persona: int
    nombre: str
    apellido: str
    documento: Optional[int]

    rol: RolPersonaTipo
    origen_rol: Literal["FICHAJE", "PLANTEL"]

    id_equipo: Optional[int]
    id_plantel: Optional[int]

    fecha_inicio: date
    fecha_fin: Optional[date]
    activo: bool

    class Config:
        from_attributes = True

# Vista personas_roles_clubes
class PersonaRolClubRead(BaseModel):
    id_persona: int
    nombre: str
    apellido: str
    rol: str | None
    estado_fichaje: str
    nombre_club: str | None

    class Config:
        from_attributes = True

class PlantelActivoIntegrante(BaseModel):
    id_plantel: int
    id_equipo: int
    nombre_plantel: str
    temporada: str
    plantel_activo: bool
    
    # Estos deben ser Optional (pueden ser None)
    id_plantel_integrante: Optional[int] = None
    id_persona: Optional[int] = None
    nombre_persona: Optional[str] = None
    apellido_persona: Optional[str] = None
    documento: Optional[int] = None
    rol_en_plantel: Optional[str] = None
    numero_camiseta: Optional[int] = None
    fecha_alta: Optional[date] = None
    fecha_baja: Optional[date] = None

    class Config:
        from_attributes = True

class PersonasArbitro(BaseModel):
    
    id_persona_rol: int
    nombre: str
    apellido: str
    documento: Optional[int] = None
    rol: RolPersonaTipo

    class Config:
        from_attributes = True

##
# Vista para tarjetas historial

class TarjetaDetalle(BaseModel):
    id_torneo: int
    torneo: str
    id_partido: int
    fecha_partido: date
    numero_fecha: int
    id_persona: int
    nombre_persona: str
    apellido_persona: str
    id_equipo: int
    equipo: str
    tipo_tarjeta: TipoTarjeta  # VERDE, AMARILLA, ROJA
    minuto: Optional[int]
    verdes: int
    amarillas: int
    rojas: int

    class Config:
        from_attributes = True

class TarjetaAcumulada(BaseModel):
    id_torneo: int
    torneo: str
    id_persona: int
    nombre_persona: str
    apellido_persona: str
    id_equipo: int
    equipo: str
    total_tarjetas: int
    total_verdes: int
    total_amarillas: int
    total_rojas: int

    class Config:
        from_attributes = True



class GoleadorTorneo(BaseModel):
    id_persona: int
    nombre: str
    apellido: str
    nombre_equipo: str
    goles_en_torneo: int
    goles_netos_en_torneo: int
    goles_totales_carrera: int
    ranking_en_torneo: int

    class Config:
        from_attributes = True

File: frontend/src/types/vistas.ts
// frontend/src/types/vistas.ts
import type { Key, ReactNode } from "react"
import type { TipoGenero, TipoRolPersona } from "../constants/enums"



export interface PlantelActivoIntegrante {
  id_plantel: number;
  id_equipo: number;
  nombre_plantel: string;
  temporada: string;
  plantel_activo: boolean;

  // Estos campos son opcionales porque el plantel puede estar vacío
  id_persona?: number | null;
  nombre_persona?: string | null;
  apellido_persona?: string | null;
  documento?: number | null;
  rol_en_plantel?: string | null;
  numero_camiseta?: number | null;
  id_plantel_integrante?: number | null;
  rol_en_palntel?: string | null;
  fecha_alta?: string | null; 
  fecha_baja?: string | null;
}


// lista de personas con roles completa
export interface PersonasRolesVista {
  rol: ReactNode
  id_persona_rol: Key | null | undefined
  id_persona: number
  nombre: string
  apellido: string
  id_rol: number
  rol_codigo: TipoRolPersona
  genero: TipoGenero   // 👈 esta línea
  nombre_club?: string | null
  fecha_desde?: string | null
  fecha_hasta?: string | null
}

// Ver una persona con sus roles asociados utiliza la vista
export interface PersonaConRoles {
  id_persona: number
  nombre: string
  apellido: string
  roles: PersonasRolesVista[]
}


export interface PersonasArbitro {

  id_persona_rol: number
  nombre: string
  apellido: string
  documento?: number | null
  rol: TipoRolPersona

}


export interface FilaPosiciones {
  id_equipo: number;
  equipo: string; // Antes era nombre_equipo
  partidos_jugados: number; // Antes era pj
  ganados: number; // Antes era pg
  empatados: number; // Antes era pe
  perdidos: number; // Antes era pp
  goles_a_favor: number; // Antes era gf
  goles_en_contra: number; // Antes era gc
  diferencia_gol: number; // Antes era dif
  puntos: number;
}

export interface TarjetaAcumulada {

  id_torneo: number;
  torneo: string;
  id_persona: number;
  nombre_persona: string;
  apellido_persona: string;
  id_equipo: number;
  equipo: string;
  total_tarjetas: number;
  total_verdes: number;
  total_amarillas: number;
  total_rojas: number;

}

export interface GoleadorTorneo {
  id_persona: number;
  nombre: string;
  apellido: string;
  nombre_equipo: string;
  goles_en_torneo: number;
  goles_netos_en_torneo: number;
  goles_totales_carrera: number;
  ranking_en_torneo: number;
}